File: src/main/java/com/lessmarkup/userinterface/model/configuration/NodeSettingsModel.java
package com.lessmarkup.userinterface.model.configuration;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.lessmarkup.Constants;
import com.lessmarkup.TextIds;
import com.lessmarkup.dataobjects.Node;
import com.lessmarkup.framework.helpers.DependencyResolver;
import com.lessmarkup.framework.helpers.JsonSerializer;
import com.lessmarkup.framework.helpers.LanguageHelper;
import com.lessmarkup.framework.helpers.LoggingHelper;
import com.lessmarkup.framework.helpers.StringHelper;
import com.lessmarkup.interfaces.cache.DataCache;
import com.lessmarkup.interfaces.cache.EntityChangeType;
import com.lessmarkup.interfaces.data.ChangeTracker;
import com.lessmarkup.interfaces.data.DomainModel;
import com.lessmarkup.interfaces.data.DomainModelProvider;
import com.lessmarkup.interfaces.exceptions.RecordValidationException;
import com.lessmarkup.interfaces.module.ModuleConfiguration;
import com.lessmarkup.interfaces.module.ModuleProvider;
import com.lessmarkup.interfaces.recordmodel.EnumSource;
import com.lessmarkup.interfaces.recordmodel.InputField;
import com.lessmarkup.interfaces.recordmodel.InputFieldType;
import com.lessmarkup.interfaces.recordmodel.InputSource;
import com.lessmarkup.interfaces.recordmodel.RecordModel;
import com.lessmarkup.interfaces.recordmodel.RecordModelCache;
import com.lessmarkup.interfaces.recordmodel.RecordModelDefinition;
import com.lessmarkup.interfaces.structure.NodeHandler;
import com.lessmarkup.interfaces.structure.Tuple;
import com.lessmarkup.userinterface.nodehandlers.configuration.NodeListNodeHandler;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.OptionalLong;
import java.util.Set;
import java.util.logging.Level;

@Component
@Scope("prototype")
public class NodeSettingsModel extends RecordModel<NodeSettingsModel> implements InputSource {

    private final ModuleProvider moduleProvider;
    private final DomainModelProvider domainModelProvider;
    private final DataCache dataCache;
    private final ChangeTracker changeTracker;
    
    private String title;
    private String handlerId;
    private Object settings;
    private String settingsModelId;
    private long nodeId;
    private boolean customizable;
    private boolean enabled;
    private boolean addToMenu;
    private String path;
    private int position;
    private String roleText;
    private final List<NodeSettingsModel> children = new ArrayList<>();
    private OptionalLong parentId;

    @Autowired
    public NodeSettingsModel(ModuleProvider moduleProvider, DomainModelProvider domainModelProvider, DataCache dataCache, ChangeTracker changeTracker) {
        super(TextIds.NODE_SETTINGS);
        this.moduleProvider = moduleProvider;
        this.domainModelProvider = domainModelProvider;
        this.dataCache = dataCache;
        this.changeTracker = changeTracker;
    }
    
    @InputField(type = InputFieldType.TEXT, textId = TextIds.TITLE, required = true)
    public void setTitle(String title) { this.title = title; }
    public String getTitle() { return title; }

    @InputField(type = InputFieldType.SELECT, textId = TextIds.HANDLER, required = true)
    public void setHandlerId(String handlerId) { this.handlerId = handlerId; }
    public String getHandlerId() { return handlerId; }

    public void setSettings(Object settings) { this.settings = settings; }
    public Object getSettings() { return settings; }

    public void setSettingsModelId(String settingsModelId) { this.settingsModelId = settingsModelId; }
    public String getSettingsModelId() { return settingsModelId; }

    public void setNodeId(long nodeId) { this.nodeId = nodeId; }
    public long getNodeId() { return nodeId; }

    public void setCustomizable(boolean customizable) { this.customizable = customizable; }
    public boolean isCustomizable() { return customizable; }

    @InputField(type = InputFieldType.CHECK_BOX, textId = TextIds.ENABLED, defaultValue = "true")
    public void setEnabled(boolean enabled) { this.enabled = enabled; }
    public boolean isEnabled() { return enabled; }

    @InputField(type = InputFieldType.CHECK_BOX, textId = TextIds.ADD_TO_MENU, defaultValue = "false")
    public void setAddToMenu(boolean addToMenu) { this.addToMenu = addToMenu; }
    public boolean isAddToMenu() { return addToMenu; }

    @InputField(type = InputFieldType.TEXT, textId = TextIds.PATH, required = true)
    public void setPath(String path) { this.path = path; }
    public String getPath() { return path; }

    public void setPosition(int order) { this.position = order; }
    public int getPosition() { return position; }

    public void setRoleText(String roleText) { this.roleText = roleText; }
    public String getRoleText() { return roleText; }

    public List<NodeSettingsModel> getChildren() { return children; }

    public void setParentId(OptionalLong parentId) { this.parentId = parentId; }
    public OptionalLong getParentId() { return parentId; }

    @Override
    public List<EnumSource> getEnumValues(String fieldName) {
        switch (StringHelper.toJsonCase(fieldName)) {
            case "handlerId": {
                List<String> modules = new ArrayList<>();
                moduleProvider.getModules().stream().map(ModuleConfiguration::getModuleType).forEach(modules::add);
                List<EnumSource> ret = new ArrayList<>();
                moduleProvider.getNodeHandlers().forEach(id -> {
                    Tuple<Class<? extends NodeHandler>, String> handler = moduleProvider.getNodeHandler(id);
                    if (!modules.contains(handler.getValue2())) {
                        return;
                    }
                    EnumSource enumSource = new EnumSource(id, NodeListNodeHandler.getHandlerName(handler.getValue1(), handler.getValue2()));
                    ret.add(enumSource);
                });
                return ret;
            }
            default:
                throw new IllegalArgumentException();
        }
    }

    public Object createNode() {
        RecordModelCache modelCache = dataCache.get(RecordModelCache.class);
        RecordModelDefinition definition = modelCache.getDefinition(NodeSettingsModel.class);
        try {
            definition.validateInput(JsonSerializer.serializeToTree(this), true, null);
        } catch (RecordValidationException ex) {
            LoggingHelper.getLogger(getClass()).log(Level.SEVERE, null, ex);
            return null;
        }

        Node target = new Node();
        target.setEnabled(enabled);
        target.setHandlerId(handlerId);
        target.setParentId(parentId);
        target.setPosition(position);
        target.setPath(path);
        target.setAddToMenu(addToMenu);
        target.setSettings(settings != null ? JsonSerializer.serialize(settings) : null);
        target.setTitle(title);

        try (DomainModel domainModel = domainModelProvider.createWithTransaction()) {
            domainModel.create(target);
            changeTracker.addChange(Node.class, target, EntityChangeType.ADDED, domainModel);
            domainModel.completeTransaction();
        }

        nodeId = target.getId();

        NodeHandler handler = DependencyResolver.resolve(moduleProvider.getNodeHandler(handlerId).getValue1());

        customizable = handler.getSettingsModel() != null;

        if (customizable) {
            settingsModelId = modelCache.getDefinition(handler.getSettingsModel()).getId();
        }

        return this;
    }
    
    public Object updateNode() {
        RecordModelCache modelCache = dataCache.get(RecordModelCache.class);
        RecordModelDefinition definition = modelCache.getDefinition(NodeSettingsModel.class);
        try {
            definition.validateInput(JsonSerializer.serializeToTree(this), false, null);
        } catch (RecordValidationException ex) {
            LoggingHelper.getLogger(getClass()).log(Level.SEVERE, null, ex);
        }

        try (DomainModel domainModel = domainModelProvider.createWithTransaction())
        {
            Node record = domainModel.query().find(Node.class, nodeId);

            record.setTitle(title);
            record.setParentId(parentId);
            record.setPath(path);
            record.setEnabled(enabled);
            record.setAddToMenu(addToMenu);

            domainModel.update(record);
            changeTracker.addChange(Node.class, record, EntityChangeType.UPDATED, domainModel);

            domainModel.completeTransaction();
        }

        return this;
    }

    public Object deleteNode() {
        try (DomainModel domainModel = domainModelProvider.createWithTransaction()) {
            Node node = domainModel.query().find(Node.class, nodeId);
            OptionalLong parentId = node.getParentId();
            domainModel.delete(Node.class, node.getId());
            changeTracker.addChange(Node.class, node, EntityChangeType.REMOVED, domainModel);

            List<Node> nodes = domainModel.query().from(Node.class).where("parentId = $", parentId).toList(Node.class);

            for (int i = 0; i < nodes.size(); i++) {
                if (nodes.get(i).getPosition() != i) {
                    nodes.get(i).setPosition(i);
                    domainModel.update(nodes.get(i));
                    changeTracker.addChange(Node.class, nodes.get(i), EntityChangeType.UPDATED, domainModel);
                }
            }

            domainModel.completeTransaction();
        }

        return null;
    }

    private void normalizeTree(List<NodeSettingsModel> nodes, NodeSettingsModel rootNode, DomainModel domainModel, Set<Long> changedNodes) {
        nodes.forEach(node -> node.getChildren().sort((n1, n2) -> Integer.compare(n1.getPosition(), n2.getPosition())));
        
        if (rootNode != null) {
            nodes.stream().filter(n -> n.getNodeId() != rootNode.getNodeId() && !n.getParentId().isPresent()).forEach(node -> {
                rootNode.getChildren().add(node);
                node.setParentId(OptionalLong.of(rootNode.getNodeId()));
                Node record = domainModel.query().find(Node.class, node.getNodeId());
                if (record.getParentId() != node.getParentId()) {
                    record.setParentId(node.getParentId());
                    domainModel.update(record);
                }

                changedNodes.add(node.getNodeId());
            });
        }

        for (NodeSettingsModel node : nodes) {
            for (int i = 0; i < node.getChildren().size(); i++) {
                NodeSettingsModel child = node.getChildren().get(i);

                if (child.getPosition() != i) {
                    child.setPosition(i);
                    Node record = domainModel.query().find(Node.class, child.getNodeId());
                    if (record.getPosition() != i) {
                        record.setPosition(i);
                        domainModel.update(record);
                    }

                    changedNodes.add(record.getId());
                }
            }
        }
    }

    public NodeSettingsModel getRootNode() {
        NodeSettingsModel rootNode;
        RecordModelCache modelCache = dataCache.get(RecordModelCache.class);
        List<NodeSettingsModel> nodes = new ArrayList<>();
        Map<Long, NodeSettingsModel> nodeIds = new HashMap<>();

        try (DomainModel domainModel = domainModelProvider.create())
        {
            domainModel.query()
                    .from(Node.class)
                    .toList(Node.class)
                    .stream()
                    .sorted((n1, n2) -> Integer.compare(n1.getPosition(), n2.getPosition()))
                    .forEach(source -> {
                NodeSettingsModel node = DependencyResolver.resolve(NodeSettingsModel.class);

                NodeHandler handler = null;

                if (source.getHandlerId() != null) {
                    Tuple<Class<? extends NodeHandler>, String> handlerReference = moduleProvider.getNodeHandler(source.getHandlerId());
                    if (handlerReference != null) {
                        handler = DependencyResolver.resolve(handlerReference.getValue1());
                    }
                }

                node.setParentId(source.getParentId());
                node.setEnabled(source.isEnabled());
                node.setHandlerId(source.getHandlerId());
                node.setNodeId(source.getId());
                node.setAddToMenu(source.isAddToMenu());
                node.setPosition(source.getPosition());
                
                if (handler != null && handler.getSettingsModel() != null) {
                    if (source.getSettings() != null) {
                        node.setSettings(JsonSerializer.deserialize(handler.getSettingsModel(), source.getSettings()));
                    }
                    node.setSettingsModelId(modelCache.getDefinition(handler.getSettingsModel()).getId());
                }
                node.setTitle(source.getTitle());
                node.setPath(source.getPath());
                node.setCustomizable(node.getSettingsModelId() != null);

                nodes.add(node);
                nodeIds.put(node.getNodeId(), node);
            });

            Set<Long> changedNodes = new HashSet<>();

            for (NodeSettingsModel node : nodes) {
                if (!node.getParentId().isPresent()) {
                    continue;
                }

                NodeSettingsModel parent = nodeIds.get(node.getParentId().getAsLong());
                if (parent == null) {
                    node.setParentId(null);
                    Node record = domainModel.query().find(Node.class, node.getNodeId());
                    record.setParentId(null);
                    domainModel.update(record);
                    changedNodes.add(node.getNodeId());
                }
                else {
                    parent.getChildren().add(node);
                }
            }
            
            Optional<NodeSettingsModel> r = nodes.stream().filter(n -> !n.getParentId().isPresent()).findFirst();
            rootNode = r.isPresent() ? r.get() : null;

            normalizeTree(nodes, rootNode, domainModel, changedNodes);

            if (changedNodes.size() > 0) {
                changedNodes.stream().forEach(nodeId -> changeTracker.addChange(Node.class, nodeId, EntityChangeType.UPDATED, domainModel));
            }
        }

        return rootNode;
    }


    public Object changeSettings() {
        try (DomainModel domainModel = domainModelProvider.createWithTransaction()) {
            Node node = domainModel.query().find(Node.class, nodeId);

            node.setSettings(settings != null ? JsonSerializer.serialize(settings) : null);

            domainModel.update(node);
            changeTracker.addChange(Node.class, node, EntityChangeType.UPDATED, domainModel);

            domainModel.completeTransaction();
        }

        return settings;
    }

    public JsonElement updateParent() {
        try (DomainModel domainModel = domainModelProvider.createWithTransaction()) {
            Node node = domainModel.query().find(Node.class, nodeId);

            Set<Long> changedNodes = new HashSet<>();

            if (node.getParentId().isPresent() != parentId.isPresent()) {
                Node newRootNode;
                Node oldRootNode;

                if (node.getParentId().isPresent()) {
                    newRootNode = node;
                    oldRootNode = domainModel.query().from(Node.class).where("parentId IS NULL").first(Node.class);
                }
                else {
                    oldRootNode = node;
                    newRootNode = domainModel.query().find(Node.class, parentId.getAsLong());
                }

                if (!newRootNode.getParentId().isPresent()) {
                    throw new IllegalArgumentException(LanguageHelper.getText(Constants.ModuleType.MAIN, TextIds.CANNOT_HAVE_TWO_ROOT_NODES));
                }

                for (Node neighbor : domainModel.query()
                        .from(Node.class)
                        .where("parentId = $ AND position > $", newRootNode.getParentId().getAsLong(), newRootNode.getPosition())
                        .toList(Node.class)) {
                    neighbor.setPosition(neighbor.getPosition() - 1);
                    domainModel.update(neighbor);
                    changedNodes.add(neighbor.getId());
                }

                newRootNode.setPosition(0);
                newRootNode.setParentId(null);
                changedNodes.add(newRootNode.getId());

                for (Node neighbor : domainModel.query()
                        .from(Node.class)
                        .where("parentId = $ AND " + Constants.Data.ID_PROPERTY_NAME + " != $ AND position > 0", newRootNode.getId(), nodeId)
                        .toList(Node.class)) {
                    neighbor.setPosition(neighbor.getPosition() + 1);
                    domainModel.update(neighbor);
                    changedNodes.add(neighbor.getId());
                }

                oldRootNode.setPosition(0);
                oldRootNode.setParentId(OptionalLong.of(newRootNode.getId()));
                changedNodes.add(oldRootNode.getId());
            }
            else if (node.getParentId().isPresent()) {
                if (node.getParentId() == parentId) {
                    if (position > node.getPosition()) {
                        for (Node neighbor : domainModel.query()
                                .from(Node.class)
                                .where("parentId = $ AND " + Constants.Data.ID_PROPERTY_NAME + " != $ AND position > $ AND position <= $", parentId, nodeId, node.getPosition(), position)
                                .toList(Node.class)) {
                            neighbor.setPosition(neighbor.getPosition() - 1);
                            domainModel.update(neighbor);
                            changedNodes.add(neighbor.getId());
                        }
                    }
                    else if (position < node.getPosition()) {
                        for (Node neighbor : domainModel.query()
                                .from(Node.class)
                                .where("parentId = $ AND " + Constants.Data.ID_PROPERTY_NAME + " != $ AND position >= $ AND position < $", parentId, nodeId, position, node.getPosition())
                                .toList(Node.class)) {
                            neighbor.setPosition(neighbor.getPosition() + 1);
                            domainModel.update(neighbor);
                            changedNodes.add(neighbor.getId());
                        }
                    }
                }
                else {
                    for (Node neighbor : domainModel.query()
                            .from(Node.class)
                            .where("parentId = $ AND position > $ AND " + Constants.Data.ID_PROPERTY_NAME + " != $", node.getParentId(), node.getPosition(), nodeId)
                            .toList(Node.class)) {
                        neighbor.setPosition(neighbor.getPosition() - 1);
                        domainModel.update(neighbor);
                        changedNodes.add(neighbor.getId());
                    }

                    for (Node neighbor : domainModel.query()
                            .from(Node.class)
                            .where("parentId = $ AND position >= $ AND " + Constants.Data.ID_PROPERTY_NAME + " != $", parentId, position, nodeId)
                            .toList(Node.class)) {
                        neighbor.setPosition(neighbor.getPosition() + 1);
                        domainModel.update(neighbor);
                        changedNodes.add(neighbor.getId());
                    }
                }
            }

            node.setParentId(parentId);
            node.setPosition(position);
            changedNodes.add(nodeId);

            for (long id : changedNodes) {
                changeTracker.addChange(Node.class, id, EntityChangeType.UPDATED, domainModel);
            }

            domainModel.completeTransaction();
        }
        
        JsonElement rootNode = JsonSerializer.serializeToTree(getRootNode());

        JsonObject ret = new JsonObject();
        ret.add("Root", rootNode);
        return ret;
    }
}

File: src/main/scala/com/lessmarkup/userinterface/model/configuration/NodeSettingsModel.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.userinterface.model.configuration

import java.util.logging.Level

import com.google.gson.{JsonElement, JsonObject}
import com.google.inject.Inject
import com.lessmarkup.dataobjects.Node
import com.lessmarkup.framework.helpers.{DependencyResolver, JsonSerializer, LanguageHelper, LoggingHelper, StringHelper}
import com.lessmarkup.framework.nodehandlers.NodeHandlerConfiguration
import com.lessmarkup.interfaces.annotations.{NodeAccessType, InputField, InputFieldType}
import com.lessmarkup.interfaces.cache.{DataCache, EntityChangeType}
import com.lessmarkup.interfaces.data._
import com.lessmarkup.interfaces.exceptions.RecordValidationException
import com.lessmarkup.interfaces.module.ModuleProvider
import com.lessmarkup.interfaces.recordmodel.{EnumSource, InputSource, RecordModel, RecordModelCache, RecordModelDefinition}
import com.lessmarkup.interfaces.structure.NodeHandlerFactory
import com.lessmarkup.userinterface.nodehandlers.configuration.NodeListNodeHandler
import com.lessmarkup.{Constants, TextIds}

import scala.collection.JavaConversions._
import scala.collection.mutable

class NodeSettingsModel @Inject() (moduleProvider: ModuleProvider, domainModelProvider: DomainModelProvider,
                                   dataCache: DataCache, changeTracker: ChangeTracker)
  extends RecordModel[NodeSettingsModel](TextIds.NODE_SETTINGS) with InputSource {

  @InputField(fieldType = InputFieldType.TEXT, textId = TextIds.TITLE, required = true)
  var title: String = null
  @InputField(fieldType = InputFieldType.SELECT, textId = TextIds.HANDLER, required = true)
  var handlerId: String = null
  var settings: JsonElement = null
  var settingsModelId: Option[String] = None
  var nodeId: Long = 0L
  var customizable: Boolean = false
  @InputField(fieldType = InputFieldType.CHECK_BOX, textId = TextIds.ENABLED, defaultValue = "true")
  var enabled: Boolean = false
  @InputField(fieldType = InputFieldType.CHECK_BOX, textId = TextIds.ADD_TO_MENU, defaultValue = "false")
  var addToMenu: Boolean = false
  @InputField(fieldType = InputFieldType.TEXT, textId = TextIds.PATH, required = true)
  var path: String = null
  var position: Int = 0
  var roleText: Option[String] = None
  private final val children: mutable.MutableList[NodeSettingsModel] = new mutable.MutableList()
  var parentId: Option[Long] = None

  def sortChildren = {
    children.sortBy(_.position)
  }

  def getChildren: Seq[NodeSettingsModel] = {
    children.seq
  }

  def getEnumValues(fieldName: String): Seq[EnumSource] = {
    StringHelper.toJsonCase(fieldName) match {
      case "handlerId" =>
        val modules = moduleProvider.getModules.map(_.getModuleType)
        moduleProvider.getNodeHandlers.flatMap(id => {
          val handler = moduleProvider.getNodeHandler(id).get
          if (!modules.contains(handler._2)) {
            None
          } else {
            Option(new EnumSource(id, NodeListNodeHandler.getHandlerName(handler._1, handler._2)))
          }
        })
      case _ => throw new IllegalArgumentException
    }
  }

  def createNode: Option[AnyRef] = {
    val modelCache: RecordModelCache = dataCache.get(classOf[RecordModelCache])
    val definition: RecordModelDefinition = modelCache.getDefinition(classOf[NodeSettingsModel]).get
    try {
      definition.validateInput(JsonSerializer.serializePojoToTree(this), isNew = true)
    }
    catch {
      case ex: RecordValidationException =>
        LoggingHelper.getLogger(getClass).log(Level.SEVERE, null, ex)
        return None
    }

    val target = new Node
    target.enabled = enabled
    target.handlerId = handlerId
    target.parentId = parentId
    target.position = position
    target.path = path
    target.addToMenu = addToMenu
    target.settings = if (settings != null) Option(settings.toString) else None
    target.title = title
    target.description = ""

    val domainModel: DomainModel = domainModelProvider.createWithTransaction
    try {
      domainModel.create(target)
      changeTracker.addChange(classOf[Node], target, EntityChangeType.ADDED, domainModel)
      domainModel.completeTransaction()
    } finally {
      if (domainModel != null) domainModel.close()
    }

    nodeId = target.id
    val handlerFactory: NodeHandlerFactory = DependencyResolver(moduleProvider.getNodeHandler(handlerId).get._1)
    val configuration = new NodeHandlerConfiguration(
      objectId = None,
      settings = None,
      accessType = NodeAccessType.READ,
      path = "",
      fullPath = ""
    )
    val handler = handlerFactory.createNodeHandler(configuration)
    customizable = handler.getSettingsModel != null
    if (customizable) {
      settingsModelId = Option(modelCache.getDefinition(handler.getSettingsModel.get).get.getId)
    }
    Option(this)
  }

  def updateNode(): Option[AnyRef] = {
    val modelCache: RecordModelCache = dataCache.get(classOf[RecordModelCache])
    val definition: RecordModelDefinition = modelCache.getDefinition(classOf[NodeSettingsModel]).get
    try {
      definition.validateInput(JsonSerializer.serializePojoToTree(this), isNew = false)
    }
    catch {
      case ex: RecordValidationException =>
        LoggingHelper.getLogger(getClass).log(Level.SEVERE, null, ex)
    }
    val domainModel: DomainModel = domainModelProvider.createWithTransaction
    try {
      val record: Node = domainModel.query.find(classOf[Node], nodeId).get
      record.title = title
      record.parentId = parentId
      record.path = path
      record.enabled = enabled
      record.addToMenu = addToMenu
      domainModel.update(record)
      changeTracker.addChange(classOf[Node], record, EntityChangeType.UPDATED, domainModel)
      domainModel.completeTransaction()
    } finally {
      if (domainModel != null) domainModel.close()
    }
    Option(this)
  }

  def deleteNode(): AnyRef = {
    val domainModel: DomainModel = domainModelProvider.createWithTransaction
    try {
      val node: Node = domainModel.query.find(classOf[Node], nodeId).get
      val nodeParentId = node.parentId
      domainModel.delete(classOf[Node], node.id)
      changeTracker.addChange(classOf[Node], node, EntityChangeType.REMOVED, domainModel)
      val nodes = domainModel
        .query
        .from(classOf[Node])
        .where("parentId = $", nodeParentId)
        .toList(classOf[Node])

      nodes.zipWithIndex.filter(n => n._1.position != n._2).foreach {
        case (n, i) =>
          n.position = i
          domainModel.update(n)
          changeTracker.addChange(classOf[Node], n, EntityChangeType.UPDATED, domainModel)
      }

      domainModel.completeTransaction()
    } finally {
      if (domainModel != null) domainModel.close()
    }
    null
  }

  private def normalizeTree(nodes: Seq[NodeSettingsModel], rootNode: NodeSettingsModel, domainModel: DomainModel, changedNodes: mutable.Set[Long]) {
    nodes.foreach(_.sortChildren)
    if (rootNode != null) {
      nodes.filter(n => n.nodeId != rootNode.nodeId && n.parentId.isEmpty)
        .foreach(node => {
          rootNode.getChildren.add(node)
          node.parentId = Option(rootNode.nodeId)
          val record = domainModel.query.find(classOf[Node], node.nodeId)
          if (record.get.parentId.get != node.parentId.get) {
            record.get.parentId = node.parentId
            domainModel.update(record.get)
          }

          changedNodes.add(node.nodeId)
      })
    }

    for (node <- nodes) {
      node.getChildren.zipWithIndex.filter(n => n._1.position != n._2).foreach {
        case (child, i) =>
          child.position = i
          val record: Node = domainModel.query.find(classOf[Node], child.nodeId).get
          if (record.position != i) {
            record.position = i
            domainModel.update(record)
          }
          changedNodes.add(record.id)
      }
    }
  }

  def getRootNode: NodeSettingsModel = {
    var rootNode: NodeSettingsModel = null
    val modelCache: RecordModelCache = dataCache.get(classOf[RecordModelCache])
    val domainModel: DomainModel = domainModelProvider.create
    try {

      val nodes: Seq[NodeSettingsModel] = domainModel.query.from(classOf[Node])
        .toList(classOf[Node])
        .sortBy(_.position)
        .map(source => {

        val node = DependencyResolver(classOf[NodeSettingsModel])

        val handlerReference = moduleProvider.getNodeHandler(source.handlerId)
        val handler =
          if (handlerReference.isDefined) {
            val factory = DependencyResolver(handlerReference.get._1)

            val configuration = new NodeHandlerConfiguration(
              objectId = None,
              settings = None,
              accessType = NodeAccessType.READ,
              path = "",
              fullPath = ""
            )

            Option(factory.createNodeHandler(configuration))
          }
          else None

        node.parentId = source.parentId
        node.enabled = source.enabled
        node.handlerId = source.handlerId
        node.nodeId = source.id
        node.addToMenu = source.addToMenu
        node.position = source.position

        if (handler.isDefined && handler.get.getSettingsModel.isDefined) {
          if (source.settings.isDefined) {
            node.settings = JsonSerializer.deserializeToTree(source.settings.get)
          }
          node.settingsModelId = Option(modelCache.getDefinition(handler.get.getSettingsModel.get).get.getId)
        }
        node.title = source.title
        node.path = source.path
        node.customizable = node.settingsModelId.isDefined

        node
      })

      val nodeIds = nodes.map(n => (n.nodeId, n)).toMap

      val changedNodes: mutable.Set[Long] = mutable.Set()

      for (node <- nodes if node.parentId.isDefined) {

        val parent = nodeIds.get(node.parentId.get)

        if (parent.isEmpty) {
          node.parentId = None
          val record = domainModel.query.find(classOf[Node], node.nodeId).get
          record.parentId = None
          domainModel.update(record)
          changedNodes.add(node.nodeId)
        }
        else {
          parent.get.getChildren.add(node)
        }
      }
      val r = nodes.find(_.parentId.isEmpty)
      rootNode = if (r.isDefined) r.get else null
      normalizeTree(nodes, rootNode, domainModel, changedNodes)
      changedNodes.foreach(id => changeTracker.addChange(classOf[Node], id, EntityChangeType.UPDATED, domainModel) )
    } finally {
      if (domainModel != null) domainModel.close()
    }
    rootNode
  }

  def changeSettings: AnyRef = {
    val domainModel: DomainModel = domainModelProvider.createWithTransaction
    try {
      val node: Node = domainModel.query.find(classOf[Node], nodeId).get
      node.settings = if (settings != null) settings.toString else null
      domainModel.update(node)
      changeTracker.addChange(classOf[Node], node, EntityChangeType.UPDATED, domainModel)
      domainModel.completeTransaction()
    } finally {
      if (domainModel != null) domainModel.close()
    }
    settings
  }

  def updateParent(): JsonElement = {
    val domainModel: DomainModel = domainModelProvider.createWithTransaction
    try {
      val node: Node = domainModel.query.find(classOf[Node], nodeId).get
      val changedNodes: mutable.Set[Long] = mutable.Set()
      if (node.parentId.isDefined != parentId.isDefined) {
        var newRootNode: Node = null
        var oldRootNode: Node = null
        if (node.parentId.isDefined) {
          newRootNode = node
          oldRootNode = domainModel.query.from(classOf[Node]).where("parentId IS NULL").first(classOf[Node], None).get
        }
        else {
          oldRootNode = node
          newRootNode = domainModel.query.find(classOf[Node], parentId.get).get
        }
        if (!newRootNode.parentId.isDefined) {
          throw new IllegalArgumentException(LanguageHelper.getText(Constants.ModuleTypeMain, TextIds.CANNOT_HAVE_TWO_ROOT_NODES))
        }
        for (neighbor <- domainModel.query.from(classOf[Node])
            .where("parentId = $ AND position > $", newRootNode.parentId.get, newRootNode.position)
            .toList(classOf[Node])) {
          neighbor.position = neighbor.position - 1
          domainModel.update(neighbor)
          changedNodes.add(neighbor.id)
        }
        newRootNode.position = 0
        newRootNode.parentId = None
        changedNodes.add(newRootNode.id)
        for (neighbor <- domainModel.query.from(classOf[Node])
            .where("parentId = $ AND " + Constants.DataIdPropertyName + " != $ AND position > 0", newRootNode.id, nodeId)
            .toList(classOf[Node])) {
          neighbor.position = neighbor.position + 1
          domainModel.update(neighbor)
          changedNodes.add(neighbor.id)
        }
        oldRootNode.position = 0
        oldRootNode.parentId = newRootNode.id
        changedNodes.add(oldRootNode.id)
      }
      else if (node.parentId.isDefined) {
        if (node.parentId.get == parentId.get) {
          if (position > node.position) {
            for (neighbor <- domainModel.query.from(classOf[Node]).where("parentId = $ AND " + Constants.DataIdPropertyName + " != $ AND position > $ AND position <= $", parentId, nodeId, node.position, position).toList(classOf[Node])) {
              neighbor.position = neighbor.position - 1
              domainModel.update(neighbor)
              changedNodes.add(neighbor.id)
            }
          }
          else if (position < node.position) {
            for (neighbor <- domainModel.query.from(classOf[Node]).where("parentId = $ AND " + Constants.DataIdPropertyName + " != $ AND position >= $ AND position < $", parentId, nodeId, position, node.position).toList(classOf[Node])) {
              neighbor.position = neighbor.position + 1
              domainModel.update(neighbor)
              changedNodes.add(neighbor.id)
            }
          }
        }
        else {
          for (neighbor <- domainModel.query.from(classOf[Node]).where("parentId = $ AND position > $ AND " + Constants.DataIdPropertyName + " != $", node.parentId, node.position, nodeId).toList(classOf[Node])) {
            neighbor.position = neighbor.position - 1
            domainModel.update(neighbor)
            changedNodes.add(neighbor.id)
          }
          for (neighbor <- domainModel.query.from(classOf[Node]).where("parentId = $ AND position >= $ AND " + Constants.DataIdPropertyName + " != $", parentId, position, nodeId).toList(classOf[Node])) {
            neighbor.position = neighbor.position + 1
            domainModel.update(neighbor)
            changedNodes.add(neighbor.id)
          }
        }
      }
      node.parentId = parentId
      node.position = position
      changedNodes.add(nodeId)
      for (id <- changedNodes) {
        changeTracker.addChange(classOf[Node], id, EntityChangeType.UPDATED, domainModel)
      }
      domainModel.completeTransaction()
    } finally {
      if (domainModel != null) domainModel.close()
    }
    val rootNode: JsonElement = JsonSerializer.serializePojoToTree(getRootNode)
    val ret: JsonObject = new JsonObject
    ret.add("Root", rootNode)
    ret
  }
}